### PR TITLE
String explosion

### PR DESCRIPTION
**Pull request by bradallred**
This could use more testing/review, and hopefully we have some console commands to help there.

This is some bug fixes/enhancements to some of our string algorithms as well cutting over some C string manipulation to use those algorithms instead so we can be better positioned to finish #1516 at some point.

### CODE REVIEW

File: gemrb/core/InterfaceConfig.cpp [hunk 1]
```diff
@@ -39,28 +39,17 @@ InterfaceConfig::~InterfaceConfig() noexcept
...
-	const char* value = NULL;
-	if (key) {
-		// lowercase the key so that the key is not case sensitive
-		char* keyCopy = strdup(key);
-		for (char* c = keyCopy; *c != '\0'; ++c) *c = tolower(*c);
-		if (configVars->get(keyCopy)) {
-			value = configVars->get(keyCopy)->c_str();
-		}
-		free(keyCopy);
+	if (configVars->get(key)) {
```
Comment (lynxlynxlynx) [inline gemrb/core/InterfaceConfig.cpp:49]: This check is redundant, the function can be simplified to the one return.

Comment (lynxlynxlynx) [reply gemrb/core/InterfaceConfig.cpp:49]: Or better yet, should check `key` as StringMap's `get` doesn't check itself.

File: gemrb/core/GameScript/GameScript.cpp [hunk 2]
```diff
@@ -1821,8 +1819,10 @@ Script* GameScript::CacheScript(const ResRef& resRef, bool AIScript)
 	if (!stream) {
 		return NULL;
 	}
-	stream->ReadLine( line, 10 );
-	if (strncmp( line, "SC", 2 ) != 0) {
+	
+	std::string line;
+	stream->ReadLine(line, 10);
+	if (line.compare(0, std::string::npos, "SC") != 0) {
```
Comment (lynxlynxlynx) [inline gemrb/core/GameScript/GameScript.cpp:1825]: Why not just use operator== in such cases?

File: gemrb/core/Streams/DataStream.cpp [hunk 3]
```diff
@@ -123,35 +123,28 @@ strret_t DataStream::ReadRegion(Region& r)
...
 	//Windows: \r\n
 	//Old Mac: \r
 	//otherOS: \n
-	while (i < ( maxlen - 1 )) {
+	while (Pos < size && i < (maxlen - 1)) {
 		char ch;
-		Read(&ch, 1);
+		i += Read(&ch, 1);
 		if (ch == '\n')
 			break;
```
Comment (lynxlynxlynx) [inline gemrb/core/Streams/DataStream.cpp:142]: Not sure if it matters, but this case now returns a value 1 greater than before.

Comment (bradallred) [reply gemrb/core/Streams/DataStream.cpp:142]: yeah, it should be fine. I think we only check it for `Error` and use `string::length()` for the actual data examination. It should be returning the number of bytes read which will always be longer.

File: gemrb/plugins/2DAImporter/2DAImporter.cpp [hunk 4]
```diff
@@ -25,81 +25,55 @@
...
 	return stricmp(str.c_str(), key.c_str()) == 0;
 }
 
-p2DAImporter::p2DAImporter() noexcept
-{
-	colNames.reserve(10);
-	rowNames.reserve(10);
-	rows.reserve(10);
-}
-
```
Comment (lynxlynxlynx) [inline gemrb/plugins/2DAImporter/2DAImporter.cpp:41]: Why did you remove these? They were added to reduce the amount of reallocations and helped significantly when you looked at a heap profile.

File: gemrb/core/GameScript/GSUtils.cpp [hunk 5]
```diff
@@ -1603,18 +1603,14 @@ static int ParseIntParam(const char *&src, const char *&str)
...
-				++cur;
-			}
+		ResRef idsTabName;
+		char *cur = idsTabName.begin();
+		const char *end = idsTabName.bufend();
+		while (cur != end && *str != ',' && *str != ')') {
+			*cur = *str;
+			++cur;
 			++str;
 		}
```
Comment (lynxlynxlynx) [inline gemrb/core/GameScript/GSUtils.cpp:1613]: This can be further simplified by using `FindNotOf` instead of the loop.

Comment (bradallred) [reply gemrb/core/GameScript/GSUtils.cpp:1613]: `FindNotOf` wont work on a c string tho. Needs further refactoring to get there.

File: gemrb/core/GameScript/GameScript.cpp (hunk 2 above)
Comment (bradallred) [reply gemrb/core/GameScript/GameScript.cpp:1825]: ah, yes, an oversight on my part certainly.

File: gemrb/plugins/2DAImporter/2DAImporter.cpp (hunk 4 above)
Comment (bradallred) [reply gemrb/plugins/2DAImporter/2DAImporter.cpp:41]: they've been added to `p2DAImporter::Open`

File: gemrb/core/Interface.cpp [hunk 6]
```diff
@@ -3616,16 +3613,18 @@ bool Interface::ResolveRandomItem(CREItem *itm) const
...
+		if (p != ResRef::npos) {
+			NewItem[p] = '\0';
+			diceSides = strtosigned<int>(NewItem.begin() + p + 1, nullptr, 10);
 		} else {
 			diceSides = 1;
 		}
-		int diceThrows = strtosigned<int>(NewItem, &endptr, 10);
+		
+		char *endptr;
+		int diceThrows = strtosigned<int>(NewItem.begin(), &endptr, 10);
```
[the remark this answers is not in the document]
Comment (bradallred) [reply gemrb/core/Interface.cpp]: I think I've done what you asked.

[the remark this answers is not in the document]
Comment (lynxlynxlynx) [reply gemrb/core/Interface.cpp]: It looks fine, thanks. It did made me look at the function more deeply and we do have at least one problem, but it's not related to these changes: it looks like we're always generating just 1 gp and the bolt example would create gold instead. I'll take a look once this is merged.

File: gemrb/core/Scriptable/Actor.cpp [hunk 7]
```diff
@@ -2023,19 +2023,15 @@ static void InitActorTables()
...
-			char **classnames = (char **) calloc(tmpbits, sizeof(char *));
-			classnames[0] = strtok(strdup(classname.c_str()), "_");
-			while (numfound<tmpbits && (classnames[numfound] = strdup(strtok(NULL, "_")))) {
-				numfound++;
-			}
-			numfound = 0;
+			auto classnames = Explode(classname, '_', tmpbits);
 			bool foundwarrior = false;
+			//we have to account for dual-swap in the multiclass field
+			auto numfound = classnames.size() + 1;
```
Comment (lynxlynxlynx) [inline gemrb/core/Scriptable/Actor.cpp:2033]: No, this should remain 0, or it will break the iteration below.

**Comment by lynxlynxlynx** [comment]
LGTM now. :)
